### PR TITLE
Improve openblas CMake logic, add generic blas option.

### PR DESCRIPTION
**Pull request by larroy**
## Description ##
Improve CMake blas logic making it consistent and add option to link with libblas. (Netlib)

### CODE REVIEW

**Comment by larroy** [comment]
Needs https://github.com/dmlc/mshadow/pull/376

**Comment by larroy** [comment]
@TaoLv @pengzhao-intel

**Comment by larroy** [comment]
@marcoabreu add [pr-awaiting-review]

**Comment by larroy** [comment]
@mxnet-label-bot add [pr-awaiting-review, build]

File: 3rdparty/mshadow/cmake/mshadow.cmake [hunk 1]
```diff
@@ -1,34 +1,40 @@
...
-elseif(BLAS STREQUAL "Open" OR BLAS STREQUAL "open")
+elseif(BLAS STREQUAL "openblas")
   find_package(OpenBLAS REQUIRED)
   include_directories(SYSTEM ${OpenBLAS_INCLUDE_DIR})
   list(APPEND mshadow_LINKER_LIBS ${OpenBLAS_LIB})
   add_definitions(-DMSHADOW_USE_CBLAS=1)
   add_definitions(-DMSHADOW_USE_MKL=0)
-elseif(BLAS STREQUAL "MKL" OR BLAS STREQUAL "mkl")
+elseif(BLAS STREQUAL "blas")
+  set(BLA_VENDOR "Generic")
```
Comment (ChaiBapchya) [inline 3rdparty/mshadow/cmake/mshadow.cmake]: Where is `BLA_VENDOR` used?
Also can we make it `BLAS_VENDOR` instead to keep it consistent?

Comment (larroy) [reply 3rdparty/mshadow/cmake/mshadow.cmake]: Can't do anything about it It's used by the CMake provided find_package.

https://cmake.org/cmake/help/v3.12/module/FindBLAS.html

**Comment by larroy** [comment]
@ChaiBapchya sure.

**Comment by larroy** [comment]
https://github.com/apache/incubator-mxnet/pull/15739

**Comment by ChaiBapchya** [comment]
retrigger!

File: 3rdparty/mshadow/cmake/mshadow.cmake [hunk 2]
```diff
@@ -1,34 +1,40 @@
 set(mshadow_LINKER_LIBS "")
 
-set(BLAS "Open" CACHE STRING "Selected BLAS library")
-set_property(CACHE BLAS PROPERTY STRINGS "Atlas;Open;MKL")
+set(BLAS "openblas" CACHE STRING "Selected BLAS library")
```
Comment (marcoabreu) [inline 3rdparty/mshadow/cmake/mshadow.cmake]: This breaks backwards compatibility

Comment (larroy) [reply 3rdparty/mshadow/cmake/mshadow.cmake]: It makes it consistent with Makefile, that's the whole point of the PR. Please reconsider your veto.

**Comment by szha** [comment]
What does a generic blas library mean? At the time of linking, it would still have to be in one of the existing blas implementations. Would it be more beneficial to capture the detected blas instead?

**Comment by larroy** [comment]
There's openblas, and libblas from netlib. Two different libraries.

File: 3rdparty/mshadow/cmake/mshadow.cmake (hunk 2 above)
Comment (marcoabreu) [reply 3rdparty/mshadow/cmake/mshadow.cmake]: None the less, it's not Backwards compatible since it changes the input arguments, right?

Comment (larroy) [reply 3rdparty/mshadow/cmake/mshadow.cmake]: Yes, you are right. The impact should be very small.

**Comment by apeforest** [comment]
I suggest we break this PR into two:
1) improve the CMake blas logic making it consistent 
2) add a generic blas option.

For 2), is there an issue filed with current approach? is any customer requesting this change?

**Comment by larroy** [comment]
Thanks for the reviews. Unfortunately I don't have time to keep pushing this PR.

**Pull request closed without merging**